Date: Wed, 28 Apr 2004 05:14:25 +1000
From: Anton Blanchard <anton@samba.org>
To: olh@suse.de
Subject: e100/e1000 mb patch


Hi,

We need mb to order between cacheable and non cacheable accesses.
This is in reference to LTC 7055.

Longer term I will try and push for the addition of io_barrier into
Linux but this will fix things in the mean time.

Anton

--- linux-2.6.5/drivers/net/e100.c~	2004-04-28 01:44:16.768827235 +1000
+++ linux-2.6.5/drivers/net/e100.c	2004-04-28 01:44:50.709368795 +1000
@@ -822,7 +822,7 @@
 	/* Order is important otherwise we'll be in a race with h/w:
 	 * set S-bit in current first, then clear S-bit in previous. */
 	cb->command |= cpu_to_le16(cb_s);
-	wmb();
+	mb();
 	cb->prev->command &= cpu_to_le16(~cb_s);
 
 	while(nic->cb_to_send != nic->cb_to_use) {
@@ -1405,7 +1405,7 @@
 		struct rfd *prev_rfd = (struct rfd *)rx->prev->skb->data;
 		put_unaligned(cpu_to_le32(rx->dma_addr),
 			(u32 *)&prev_rfd->link);
-		wmb();
+		mb();
 		prev_rfd->command &= ~cpu_to_le16(cb_el);
 		pci_dma_sync_single_for_device(nic->pdev, rx->prev->dma_addr,
 					       sizeof(struct rfd), PCI_DMA_TODEVICE);
